src/chains/ethereum/ethereum/src/helpers/simulations.ts

```diff
@@ -0,0 +1,594 @@
+      nonce: string;
+      balance: string;
+      state: { [slot: string]: string };
+      stateDiff: never;
+    }>
+  | Partial<{
+      code: string;
+      nonce: string;
+      balance: string;
+      state: never;
+      stateDiff: { [slot: string]: string };
+    }>;
+
+export type CallOverrides = {
+  [address: string]: CallOverride;
+};
+
+export type CreateAccessListResult = {
+  accessList: AccessList;
+  gasUsed: Quantity;
+};
+
+/**
+ * Stripped down version of the type from EthereumJs
+ */
+interface VMRunCallOpts {
+  caller?: EthereumJsAddress;
+  data?: Buffer;
+  gasPrice?: BN;
+  gasLimit?: BN;
+  to?: EthereumJsAddress;
+  value?: BN;
+  block?: any;
+}
+
+type CloneVmOptions = {
+  blockchain: Blockchain;
+  common: Common;
+  simulationBlock: Block;
+};
+
+type CloneVmResult = {
+  vm: VM;
+  stateManager: DefaultStateManager;
+  stateTrie: GanacheTrie;
+};
+
+type SimulatorCommonOptions = CloneVmOptions & {
+  options: EthereumInternalOptions;
+  emitStepEvent: boolean;
+  transaction: SimulationTransaction;
+};
+
+type RunCallOptions = SimulatorCommonOptions & {
+  overrides?: CallOverrides;
+};
+
+type CreateAccessListOptions = SimulatorCommonOptions;
+
+type RunCallSetupOptions = {
+  common: Common;
+  stateTrie: GanacheTrie;
+  stateManager: DefaultStateManager;
+  transaction: SimulationTransaction;
+  overrides?: CallOverrides;
+};
+
+type RunCallSetupResult = {
+  runCallOpts: VMRunCallOpts;
+  accessListExclusions: EthereumJsAddress[];
+  addressesOnlyStorage: EthereumJsAddress[];
+  intrinsicGas: bigint;
+};
+
+type InternalRunCallOptions = {
+  blockchain: Blockchain;
+  runCallOpts: VMRunCallOpts;
+  vm: VM;
+  options: EthereumInternalOptions;
+  emitStepEvent: boolean;
+};
+
+const toLightEJSAddress = (address?: Address): EthereumJsAddress => {
+  if (address) {
+    const buf = address.toBuffer();
+    return { buf, equals: (a: { buf: Buffer }) => buf.equals(a.buf) } as any;
+  } else {
+    return null;
+  }
+};
+
+/**
+ * Calls the StateManager's `addWarmedAddress` function for the transaction
+ * caller, all precompiles, and the to address, if applicable.
+ * @param stateManager
+ * @param caller
+ * @param to
+ * @returns An array of precompile addresses
+ */
+const warmDefaults = (
+  stateManager: DefaultStateManager,
+  caller: EthereumJsAddress,
+  to?: EthereumJsAddress
+) => {
+  // handle Berlin hardfork warm storage reads
+  const precompiles: EthereumJsAddress[] = warmPrecompiles(stateManager);
+  stateManager.addWarmedAddress(caller.buf);
+  if (to) stateManager.addWarmedAddress(to.buf);
+
+  return precompiles;
+};
+
+const warmAccessList = (
+  stateManager: DefaultStateManager,
+  accessList: AccessList
+) => {
+  for (const { address, storageKeys } of accessList) {
+    const addressBuf = Address.toBuffer(address);
+    stateManager.addWarmedAddress(addressBuf);
+    for (const slot of storageKeys) {
+      stateManager.addWarmedStorage(addressBuf, Data.toBuffer(slot, 32));
```

Comment:
I check our code coverage reporter and this line isn't ever hit somehow. Can you look into this?

Reply:
This is only used when an access list is used in a transaction that is passed to `eth_call`. Ethereumjs' `runTx` function (used for regular mining) does this internally, but `runCall` (used for our simulations), does not. However, we don't want to call this when _creating_ an access list because it makes it so that the user-supplied access list is always used, even though it shouldn't be. So all of that to say, it's a pretty specific test case that is loosely related to this PR, and I'd rather leave it for another PR. Let me know if you disagree.